### ISSUE TITLE
How can I load only one or a few pages at one time?

### ISSUE THREAD

**Issue by godwin668** [mozilla/pdf.js#8897]
Even if pdf.js support Range Request that load pages lazily. But still load too many pages when open the page first time.

Is there any option to set number of pages or chunks loaded at one time?

**Comment by kuiran** [comment]
i want to know how to render all pages at one time using the demo lol

**Comment by Snuffleupagus** [comment]
> Is there any option to set number of pages or chunks loaded at one time?

Given that lack of details it's very difficult to know the context of the question, but if you're using the default viewer with `PDFJS.disableAutoFetch = true` set, the pages will only load on demand (i.e. when they become visible); refer to https://github.com/mozilla/pdf.js/blob/ba219965760d8521a7f3794184b653f93c65c5b9/src/display/global.js#L176-L186

Closing as answered.

**Comment by godwin668** [comment]
I've already set PDFJS.disableAutoFetch = true, but still load too many pages. There are several or dozens of Range Requests and loaded too many pages and waste a lot of network. I just only want to load only a few pages when page loaded first time.

**Comment by Snuffleupagus** [comment]
> There are several or dozens of Range Requests and loaded too many pages and waste a lot of network. I just only want to load only a few pages when page loaded first time.

Considering how data is structured within PDF files, it's usually not possible to *only* request the data of a particular page. Note how important data (such as the XRef table) is usually located at the end of the file, unless the PDF file itself is linearized.

This issue is unfortunately missing too many of the requested details for anyone to provide additional assistance/answers, please make sure that you consider https://github.com/mozilla/pdf.js/blob/master/.github/ISSUE_TEMPLATE.md and https://github.com/mozilla/pdf.js/blob/master/.github/CONTRIBUTING.md when opening an issue (the part about providing a working example seems very relevant here).

**Comment by lyqeyes** [comment]
I also met this problem. By executing _PDFJS.disableAutoFetch = true_ in global scope did't make any difference. 

Then I read the source code, I found it can't set disableAutoFetch property to PDFJS Instance at all.

So, my solution is modifying the source code, setting the default value of disableAutoFetch as true, in file _viewer.js_.(my version is v1.9.426.) 

With some other necessary modifications,  it worked in my server finally.

If someone need more detail, reply or contact directly.

**Comment by matiaso** [comment]
In case someone still has the issue you have to set `disableStream` to `true` as specified here:

https://github.com/mozilla/pdf.js/blob/ba219965760d8521a7f3794184b653f93c65c5b9/src/display/global.js#L181-L182

Advise to myself: always read the NOTE :)

**Comment by balinch** [comment]
@lyqeyes Hello, can you share with me your solution/viewer.js file? I have sent you an email.

**Comment by tofi1980** [comment]
Try this to lazy load pdfs:

pdfjsLib.getDocument({ url: 'http://www.xxx.yy/zz.pdf', disableAutoFetch: true, disableStream: true}).then(function(pdf)

**Comment by parmodkumarr** [comment]
<html>
<head>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script src="pdf.js"></script>
<script src="jquery.js"></script>
</head>
<body>
<h1>PDF.js 'Hello, world!' example</h1>
<div id="canvas_div"></div>
<body>
<script>
// If absolute URL from the remote server is provided, configure the CORS
// header on that server.
var url = 'pdff.pdf';          
// Loaded via <script> tag, create shortcut to access PDF.js exports.
var pdfjsLib = window['pdfjs-dist/build/pdf'];            
// The workerSrc property shall be specified.
pdfjsLib.GlobalWorkerOptions.workerSrc = 'worker.js';
var loadingTask = pdfjsLib.getDocument(url);
loadingTask.promise.then(function(pdf) {
var __TOTAL_PAGES = pdf.numPages; 
// Fetch the first page
var pageNumber = 1;			  
for( let i=1; i<=__TOTAL_PAGES; i+=1){
var id ='the-canvas'+i;
$('#canvas_div').append("<div style='background-color:gray;text-align: center;padding:20px;' ><canvas calss='the-canvas' id='"+id+"'></canvas></div>");				
var canvas = document.getElementById(id);
//var pageNumber = 1;
renderPage(canvas, pdf, pageNumber++, function pageRenderingComplete() {
if (pageNumber > pdf.numPages) {
return; 
}
// Continue rendering of the next page
renderPage(canvas, pdf, pageNumber++, pageRenderingComplete);
});				
}            	  
});           
function renderPage(canvas, pdf, pageNumber, callback) {
pdf.getPage(pageNumber).then(function(page) {
var scale = 1.5;
var viewport = page.getViewport({scale: scale});            
var pageDisplayWidth = viewport.width;
var pageDisplayHeight = viewport.height;
//var pageDivHolder = document.createElement();
// Prepare canvas using PDF page dimensions
//var canvas = document.createElement(id);
var context = canvas.getContext('2d');
canvas.width = pageDisplayWidth;
canvas.height = pageDisplayHeight;
// pageDivHolder.appendChild(canvas);           
// Render PDF page into canvas context
var renderContext = {
canvasContext: context,
viewport: viewport
};
page.render(renderContext).promise.then(callback);
});
}           
</script>
<html>

**Comment by parmodkumarr** [comment]
<html>
<head>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script src="pdf.js"></script>
<script src="jquery.js"></script>
</head>
<body>
<h1>PDF.js 'Hello, world!' example</h1>
<div id="canvas_div"></div>
<body>
<script>
// If absolute URL from the remote server is provided, configure the CORS
// header on that server.
var url = 'pdff.pdf';          
// Loaded via <script> tag, create shortcut to access PDF.js exports.
var pdfjsLib = window['pdfjs-dist/build/pdf'];            
// The workerSrc property shall be specified.
pdfjsLib.GlobalWorkerOptions.workerSrc = 'worker.js';
var loadingTask = pdfjsLib.getDocument(url);
loadingTask.promise.then(function(pdf) {
var __TOTAL_PAGES = pdf.numPages; 
// Fetch the first page
var pageNumber = 1;			  
for( let i=1; i<=__TOTAL_PAGES; i+=1){
var id ='the-canvas'+i;
$('#canvas_div').append("<div style='background-color:gray;text-align: center;padding:20px;' ><canvas calss='the-canvas' id='"+id+"'></canvas></div>");				
var canvas = document.getElementById(id);
//var pageNumber = 1;
renderPage(canvas, pdf, pageNumber++, function pageRenderingComplete() {
if (pageNumber > pdf.numPages) {
return; 
}
// Continue rendering of the next page
renderPage(canvas, pdf, pageNumber++, pageRenderingComplete);
});				
}            	  
});           
function renderPage(canvas, pdf, pageNumber, callback) {
pdf.getPage(pageNumber).then(function(page) {
var scale = 1.5;
var viewport = page.getViewport({scale: scale});            
var pageDisplayWidth = viewport.width;
var pageDisplayHeight = viewport.height;
//var pageDivHolder = document.createElement();
// Prepare canvas using PDF page dimensions
//var canvas = document.createElement(id);
var context = canvas.getContext('2d');
canvas.width = pageDisplayWidth;
canvas.height = pageDisplayHeight;
// pageDivHolder.appendChild(canvas);           
// Render PDF page into canvas context
var renderContext = {
canvasContext: context,
viewport: viewport
};
page.render(renderContext).promise.then(callback);
});
}           
</script>
<html>

**Comment by parmodkumarr** [comment]
<html>
   <head>
      <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
      <script src="pdf.js"></script>
      <script src="jquery.js"></script>
   </head>
   <body>
      <h1>PDF.js 'Hello, world!' example</h1>
      <div id="canvas_div"></div>
      <body>
         <script>
            // If absolute URL from the remote server is provided, configure the CORS
            // header on that server.
            var url = 'pdff.pdf';
            
            // Loaded via <script> tag, create shortcut to access PDF.js exports.
            var pdfjsLib = window['pdfjs-dist/build/pdf'];
            
            // The workerSrc property shall be specified.
            pdfjsLib.GlobalWorkerOptions.workerSrc = 'worker.js';
            var loadingTask = pdfjsLib.getDocument(url);
                loadingTask.promise.then(function(pdf) {
            			 var __TOTAL_PAGES = pdf.numPages; 
            			  // Fetch the first page
            			  var pageNumber = 1;			  
            			for( let i=1; i<=__TOTAL_PAGES; i+=1){
            	var id ='the-canvas'+i;
            	$('#canvas_div').append("<div style='background-color:gray;text-align: center;padding:20px;' ><canvas calss='the-canvas' id='"+id+"'></canvas></div>");				
            				  var canvas = document.getElementById(id);
            				  //var pageNumber = 1;
            				  renderPage(canvas, pdf, pageNumber++, function pageRenderingComplete() {
            					if (pageNumber > pdf.numPages) {
            					  return; 
            					}
            					// Continue rendering of the next page
            					renderPage(canvas, pdf, pageNumber++, pageRenderingComplete);
            				  });				
            			}
            	  
                });
            
                function renderPage(canvas, pdf, pageNumber, callback) {
                  pdf.getPage(pageNumber).then(function(page) {
                    var scale = 1.5;
                    var viewport = page.getViewport({scale: scale});
            
                    var pageDisplayWidth = viewport.width;
                    var pageDisplayHeight = viewport.height;
            		//var pageDivHolder = document.createElement();
                    // Prepare canvas using PDF page dimensions
                    //var canvas = document.createElement(id);
                    var context = canvas.getContext('2d');
                    canvas.width = pageDisplayWidth;
                    canvas.height = pageDisplayHeight;
                   // pageDivHolder.appendChild(canvas);
            
                    // Render PDF page into canvas context
                    var renderContext = {
                      canvasContext: context,
                      viewport: viewport
                    };
                    page.render(renderContext).promise.then(callback);
                  });
                }           
         </script>
         <html>

**Comment by Abilashinamdar** [comment]
> I also met this problem. By executing _PDFJS.disableAutoFetch = true_ in global scope did't make any difference.
> 
> Then I read the source code, I found it can't set disableAutoFetch property to PDFJS Instance at all.
> 
> So, my solution is modifying the source code, setting the default value of disableAutoFetch as true, in file _viewer.js_.(my version is v1.9.426.)
> 
> With some other necessary modifications, it worked in my server finally.
> 
> If someone need more detail, reply or contact directly.

Hi lyqeyes,

I have followed the same approach of making disableAutoFetch as true but as you said already it didn't make any difference. Along with it I have also tried of disableStream and disableRange properties but nothing worked out.
Can you please walk me through the modifications on what all you have been done. 
My requirement is like I want to load each page of PDF only on upfront I mean only when I visit that particular page.

**Comment by legistek** [comment]
> My requirement is like I want to load each page of PDF only on upfront I mean only when I visit that particular page.

I have the same goal but I don't think it's possible. By setting disablestream and disableautofetch to true, I can get it to the point where it will only load UP TO the desired page, but I can't get it to load JUST the desired page.

In other words if I have a 1,000 page linearized PDF and I set it to open page 1 only, I can prevent it from loading past 1.

But if I try to get page 1000, it downloads the entire PDF.

Would appreciate if someone could clarify for certain if this is normal/expected. I've been looking through all the issues and documentation to get clarification on this and I haven't been able to determine the answer definitively. Thanks.

**Comment by akash-agr** [comment]
Hi, 
I am using version 2.0.338 and facing the same issue. Can anyone please help me how can I solve this problem? Thanks a lot. 

@lyqeyes Sir, Can you please tell me what changes have you done to the source code?  

Thanks a lot.

**Comment by legistek** [comment]
@akash-agr Since my post last year I changed to the latest (or at least later) version - 2.9.359 - and it works properly with linearized PDFs. The two keys are: 
(1) When you open the PDF in Acrobat and check Properties it must say "Fast Web View: Yes". Note that not all PDF generators support saving in this format. For example, PDFSharp does not, but Aspose does. I use Aspose to linearize all PDFs at import time.
(2) The server must support content-range requests. Many do not by default. If you're using ASP.NET Core, use the `enableRangeProcessing` parameter for `File` or `PhysicalFile`. Or you can parse and respond to the content-range headers manually. But if you just put the PDF on a dumb HTTP server, Azure blob, or something like that, you won't get the benefits of linearization because the server cannot serve only a portion of the file at a time.

If you satisfy those two conditions, PDF.JS will ONLY download as much of the PDF as necessary to render any particular page, even if you haven't downloaded the contents prior to that page.

**Comment by akash-agr** [comment]
Hi Peter, Thanks for writing back.

1. I can also linearize the PDFs at import time. 
2. I am using react framework. Do you have any idea if anything else I need to do for react application?  
3. Do I have to necessarily use the latest version (2.9.359)?  Is not possible in v2.0.338? I switched to an older version as I couldn't figure out how to create a PDFJS instance in the latest versions. 

Many thanks, I really appreciate your help as I am stuck for many days.

**Comment by legistek** [comment]
I don't think you should need to do anything special for React. But sorry I don't know how far back proper linearization support goes. I just know that the version I was previously using (from 2017, don't know the version # sorry) didn't do it correctly and the newest one does. But also the later versions are easy to use and there's no special instantiation required. Just do (in Typescirpt):

```
var pdfjs = (window as any).pdfjsLib;
var pdf = await pdfjs.getDocument(...).promise;
var page = await pdf.getPage(...)
```
etc

**Comment by akash-agr** [comment]
Thanks a lot, Peter. This answers a lot of my questions. I will move to newer versions and try if it works after linearizing the PDFs.

**Comment by projectcss** [comment]
> > My requirement is like I want to load each page of PDF only on upfront I mean only when I visit that particular page.
> 
> I have the same goal but I don't think it's possible. By setting disablestream and disableautofetch to true, I can get it to the point where it will only load UP TO the desired page, but I can't get it to load JUST the desired page.
> 
> In other words if I have a 1,000 page linearized PDF and I set it to open page 1 only, I can prevent it from loading past 1.
> 
> But if I try to get page 1000, it downloads the entire PDF.
> 
> Would appreciate if someone could clarify for certain if this is normal/expected. I've been looking through all the issues and documentation to get clarification on this and I haven't been able to determine the answer definitively. Thanks.

May I ask if it has been resolved
